.github/workflows/tests-gpu.yml

```diff
@@ -0,0 +1,111 @@
+name: GPU Tests
+on:
+  push:
+    branches:
+      - master
+  pull_request:
+    types:
+      - opened
+      - reopened
+      - labeled
+      - synchronize
+
+concurrency:
+  group: gpu-test-${{ github.ref }}
+  cancel-in-progress: true
+
+jobs:
+  gpu-tests:
+    runs-on:
+      - self-hosted
+      - ubuntu-22.04
+      - gpu
+
+    if: github.event_name == 'push' || (github.event_name == 'pull_request' && contains(github.event.pull_request.labels.*.name, 'gpu'))
+
+    strategy:
+      max-parallel: 2
+      matrix:
+        python-version:
+          - 3.7
+          - 3.8
+          - 3.9
+          - '3.10'
+
+    env:
+      SUITE: gpu
+
+    steps:
+      - name: Update PATH and LD_LIBRARY_PATH
+        run: |
+          echo "$HOME/.local/bin" >> $GITHUB_PATH
+          echo "/usr/local/cuda/bin" >> $GITHUB_PATH
+          echo "LD_LIBRARY_PATH=/usr/local/cuda/lib64${LD_LIBRARY_PATH:+:${LD_LIBRARY_PATH}}" >> $GITHUB_ENV
+
+      - name: Checkout Repo
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 1
+
+      - name: Setup Python
+        id: setup_python
+        uses: actions/setup-python@v4
+        with:
+          python-version: ${{ matrix.python-version }}
+
+      # Since the self-hosted runner can be re-used. It is best to set up all package
+      # installations in a virtual environment that gets cleaned at the end of each workflow run
+      - name: Setup Python virtual environment
+        id: setup_venv
+        env:
+          VENV_NAME: venv_${{ steps.setup_python.outputs.python-version }}_${{ github.sha }}
+        run: |
+          # Clear any pre-existing venvs
+          rm -rf venv_*
```

Comment:
Since we use matrix above, these should be run on different runners/directories? I guess that means there is no issue with multiple concurrent jobs deleting a virtualenv in use?

Reply:
Each runner will only process one workflow at a time, so there should not be a situation where multiple concurrent jobs are deleting a venv in use, this line I put in more as a cleanup just so if there is old venvs in the `$GITHUB_WORKSPACE` then the storage is cleared up before anything else happens

But you'll see each job at the end cleans up their own venv once the unittests are run so realistically there shouldn't be virtualenvs being left from older jobs.

Reply:
Realistically we shouldn't need a line like this in this step, but I had it in just cause. But I can remove it to make things cleaner and avoid any confusion

Reply:
All good here. I just noticed the `max-parallel: 2` and was worried they'd hit the same GPU runner, but happy to go with whichever option you'd prefer.

Reply:
Oh I see, yeah the max-parallel sets how many of the matrix jobs runs at once, but each workflow still runs on one runner exclusively at a time. So in this case you'd see two of the Python versions be processed before the rest are